autometa/validation/cluster_process.py

```diff
@@ -1,221 +1,268 @@
+        elif args.kingdom == 'archaea':
+            total_markers = 162
+
+        number_markers_found = 0
+        number_unique_markers = len(markers_in_cluster[cluster])
+        for pfam in markers_in_cluster[cluster]:
+            number_markers_found += markers_in_cluster[cluster][pfam]
+
+        # The following protects for the edge case where a cluster has zero marker genes
+        if number_unique_markers == 0:
+            completness = 'unknown'
+        else:
+            completeness = (number_unique_markers / total_markers) * 100
+        if number_markers_found == 0:
+            purity = 'unknown'
+        else:
+            purity = (number_unique_markers / number_markers_found) * 100
+
+        # Calculate average GC and cov, weighted by sequence length
+        weighted_gc_av = 0.0
+        weighted_cov_av = 0.0
+        for seq_record in cluster_sequences[cluster]:
+            seq_name = str(seq_record.id)
+            seq_length = contig_info[seq_name]['length']
+            seq_gc = contig_info[seq_name]['gc']
+            seq_cov = contig_info[seq_name]['cov']
+            seq_length_frac = seq_length / total_size
+            weighted_gc_av += seq_gc * seq_length_frac
+            weighted_cov_av += seq_cov * seq_length_frac
+
+        # Write line in summary table
+        line = '\t'.join(map(str,[
+            cluster,
+            total_size,
+            longest_contig,
+            n50,
+            number_contigs,
+            completeness,
+            purity,
+            weighted_cov_av,
+            weighted_gc_av
+        ]))+'\n'
+        fh.write(line)
+
+        # Write individual fasta file
+        cluster_outfpath = os.path.join(args.outdir, f'cluster_{cluster}.fasta')
+        SeqIO.write(cluster_sequences[cluster], cluster_outfpath, 'fasta')
+
+    fh.close()
+
+    # If the user has specified --do_taxonomy, then they also need to specify --db_dir
+    if args.do_taxonomy:
+        if not args.db_dir:
+            logger.error('Error! If you want to analyze taxonomy, you need to specify a path to database files (--db_dir)')
+            exit(1)
+
+        if not os.path.isdir(args.db_dir):
+            logger.error('Error! DB dir ' + args.db_dir + ' does not exist')
+            exit(1)
+        if not os.path.isfile(args.db_dir + '/names.dmp'):
+            logger.error('Error! Cannot find names.dmp in ' + args.db_dir)
+            exit(1)
+
+        if not os.path.isfile(args.db_dir + '/nodes.dmp'):
+            logger.error('Error! Cannot find nodes.dmp in ' + args.db_dir)
+            exit(1)
+        # Now run cluster_taxonomy.py
+        taxonomy_output_path = os.path.join(args.outdir,'cluster_taxonomy.tab')
+        cmd = f'cluster_taxonomy.py -t {args.bin_table} -c {args.column} -x {args.db_dir} -o {taxonomy_output_path}'
+
+        run_command(cmd)
+
+
+def main(args):
+    logger.info(args.hello_world)
+    # operations on args.positional
+    # operations on args.optional
```

Comment:
There are two definitions of function `main`. This second one should be taken out.

Reply:
Done!